include/libcyphal/transport/transfer_id_generators.hpp

```diff
@@ -0,0 +1,141 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef LIBCYPHAL_TRANSPORT_TRANSFER_ID_GENERATORS_HPP_INCLUDED
+#define LIBCYPHAL_TRANSPORT_TRANSFER_ID_GENERATORS_HPP_INCLUDED
+
+#include "types.hpp"
+
+#include <cetl/cetl.hpp>
+#include <cetl/pf17/cetlpf.hpp>
+
+#include <bitset>
+#include <cstddef>
+#include <utility>
+
+namespace libcyphal
+{
+namespace transport
+{
+
+/// Internal implementation details of the Presentation layer.
+/// Not supposed to be used directly by the users of the library.
+///
+namespace detail
+{
+
+/// @brief Defines a trivial transfer ID generator.
+///
+/// The generator is trivial in the sense that it simply modulo increments the transfer ID.
+/// B/c modulo is expected to be quite big (like >= 2^48), collisions of transfer ids are unlikely.
+/// Normally in use for UDP transport, where the modulo is `2^64 - 1`.
+///
+class TrivialTransferIdGenerator
+{
+public:
+    explicit TrivialTransferIdGenerator(const TransferId modulo) noexcept
+        : modulo_{modulo}
+        , next_transfer_id_{0}
+    {
+        CETL_DEBUG_ASSERT(modulo > 0, "Transfer ID modulo must be greater than 0.");
+    }
+
+    /// @brief Returns the next transfer ID.
+    ///
+    CETL_NODISCARD TransferId nextTransferId() noexcept
+    {
+        return std::exchange(next_transfer_id_, (next_transfer_id_ + 1) % modulo_);
+    }
+
+    /// @brief Sets next transfer ID.
+    ///
+    /// In use for testing purposes.
+    ///
+    void setNextTransferId(const TransferId transfer_id) noexcept
+    {
+        CETL_DEBUG_ASSERT(transfer_id < modulo_, "Valid Transfer ID must be less than modulo.");
+        next_transfer_id_ = transfer_id;
+    }
+
+private:
+    // MARK: Data members:
+
+    const TransferId modulo_;
+    TransferId       next_transfer_id_;
+
+};  // TrivialTransferIdGenerator
+
+// MARK: -
+
+/// @brief Defines a small range transfer ID generator.
+///
+/// The generator tracks allocated transfer ids by using bitset operations.
+/// Its `Size` and modulo is expected to be quite small (like <= 2^8).
+/// Normally in use for CAN transport, where the modulo is `2^5`.
+///
+template <std::size_t Size>
+class SmallRangeTransferIdGenerator
+{
+    static_assert(Size > 0, "Size must be greater than 0.");
+
+public:
+    explicit SmallRangeTransferIdGenerator(const TransferId modulo) noexcept
+        : modulo_{modulo}
+        , next_transfer_id_{0}
+        , in_use_transfer_ids_{}
+    {
+        CETL_DEBUG_ASSERT(modulo > 0, "Transfer ID modulo must be greater than 0.");
+        CETL_DEBUG_ASSERT(modulo <= Size, "Transfer ID modulo must be <= than `Size`.");
+    }
+
+    /// @brief Returns the next available (not in use) transfer ID.
```

Comment:
Please document that the worst-case complexity is linear of the number of pending requests.